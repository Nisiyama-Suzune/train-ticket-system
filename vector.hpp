#ifndef SJTU_VECTOR_HPP
#define SJTU_VECTOR_HPP
#include "exceptions.hpp"
#include <iostream>
#include <climits>
#include <cstddef>
#include <QDataStream>
#include <QTextStream>


namespace sjtu {
/**
 * a data container like std::vector
 * store data in a successive memory and support random access.
 */
template<typename T>
class vector {
private:
	T * storage;
	int MaxSize;
	int NowLength;
    void DoubleSpace()
    {
        T * tem = storage;
        MaxSize *= 2;
        storage = (T*)operator new[](MaxSize * sizeof(T));
        for(int i = 0; i < NowLength; ++i)
        {
//			storage[i] = new T(tem[i]);
			new(storage + i) T(tem[i]);
            tem[i].~T();
        }
		operator delete [](tem);
    }

public:
	/**
	 * TODO
	 * a type for actions of the elements of a vector, and you should write
	 *   a class named const_iterator with same interfaces.
	 */
	/**
	 * you can see RandomAccessIterator at CppReference for help.
	 */
	class const_iterator;
	class iterator {
		friend class vector;
	private:
		int pos;
		vector<T> * vec;
 		/**
		 * TODO add data members
		 *   just add whatever you want.
		 */
	public:
		/**

		 * return a new iterator which pointer n-next elements
		 *   even if there are not enough elements, just return the answer.
		 * as well as operator-
		 */
		iterator(int obj = 0, vector<T> *p = nullptr) : pos(obj), vec(p){}
		iterator operator+(const int &n) const
		{
			return iterator(pos + n, vec);
			//TODO
		}
		iterator operator-(const int &n) const
		{
			return iterator(pos - n, vec);
			//TODO
		}
		// return th distance between two iterator,
		// if these two iterators points to different vectors, throw invaild_iterator.
		int operator-(const iterator &rhs) const
		{
			if(vec != rhs.vec) throw invalid_iterator();
			if(pos > rhs.pos) return pos - rhs.pos;
			return rhs.pos - pos;
		}
		iterator operator+=(const int &n)
		{
			pos += n;
			return *this;
			//TODO
		}
		iterator operator-=(const int &n)
		{
			pos -= n;
			return *this;
			//TODO
		}
		/**
		 * TODO iter++
		 */
		iterator operator++(int)
		{
			int obj = pos;
			++pos;
			return iterator(obj, vec);
		}
		/**
		 * TODO ++iter
		 */
		iterator& operator++()
		{
			++pos; return *this;
		}
		/**
		 * TODO iter--
		 */
		iterator operator--(int)
		{
			int obj = pos;
			++pos;
			return iterator(obj, vec);
		}
		/**
		 * TODO --iter
		 */
		iterator& operator--()
		{
			--pos;
			return *this;
		}
		/**
		 * TODO *it
		 */
		T& operator*() const
		{
			return vec -> storage[pos];
		}
		/**
		 * a operator to check whether two iterators are same (pointing to the same memory).
		 */
		bool operator==(const iterator &rhs) const
		{
			return (vec == rhs.vec && pos == rhs.pos);
		}
		/**
		 * some other operator for iterator.
		 */
		bool operator!=(const iterator &rhs) const
		{
			return (vec != rhs.vec || pos != rhs.pos);
		}
	};
	/**
	 * TODO
	 * has same function as iterator, just for a const object.
	 */
	class const_iterator
	{

		friend class vector;
	private:
		int pos;
		const vector<T> * vec;
	public:
	    const_iterator(int obj = 0,const vector<T> * p = nullptr) : pos(obj), vec(p){}
		const_iterator operator+(const int &n) const
		{
			return const_iterator(pos + n, vec);
			//TODO
		}
		const_iterator operator-(const int &n) const
		{
			return const_iterator(pos - n, vec);
			//TODO
		}
		// return th distance between two iterator,
		// if these two iterators points to different vectors, throw invaild_iterator.
		int operator-(const const_iterator &rhs) const
		{
			if(vec != rhs.vec) throw invalid_iterator();
			if(pos > rhs.pos) return pos - rhs.pos;
			return rhs.pos - pos;
			//TODO
		}
		const T& operator*() const {
			return vec -> storage[pos];
		}
		const T* operator->() const {
            return &(operator*());
		}

		bool operator==(const const_iterator &rhs) const
		{
			return (vec == rhs.vec && pos = rhs.pos);
		}
		bool operator!=(const const_iterator &rhs) const
		{
			return (vec != rhs.vec || pos != rhs.pos);
		}
		const_iterator operator++(int)
		{
			int obj = pos;
			++pos;
			return const_iterator(obj, vec);
		}
		/**
		 * TODO ++iter
		 */
		const_iterator& operator++()
		{
			++pos; return *this;
		}
		/**
		 * TODO iter--
		 */
		const_iterator operator--(int)
		{
			int obj = pos;
			--pos;
			return const_iterator(obj, vec);
		}
		/**
		 * TODO --iter
		 */
		const_iterator& operator--()
		{
			--pos; return *this;
		}


	};

	vector(int initial = 100)
	{
		storage = (T*)operator new[](initial * sizeof(T));
		MaxSize = initial;
		NowLength = 0;
	}
	vector(const vector &other)
	{
		MaxSize = other.MaxSize;
		NowLength = other.NowLength;
		storage = (T*)operator new[](MaxSize * sizeof(T));
		//std::cout << "hey" << '\n';
		for(int i = 0; i < NowLength; ++i)
		{
		    new(storage + i)T(other.storage[i]);
			//storage[i] = other.storage[i];
		}
	}

	~vector()
	{
	    for(int i = 0; i < NowLength; ++i)
        {
            (storage + i) -> ~T();
        }
		operator delete [](storage);
	}
	/**
	 * TODO Assignment operator
	 */
	vector &operator=(const vector &other)
	{
		if (this == &other)
			return *this;
		MaxSize = other.MaxSize;
		NowLength = other.NowLength;
		for(int i = 0; i < NowLength; ++i)
		{
			new(storage + i)T(other.storage[i]);
		}
		return *this;
	}
	/**
	 * assigns specified element with bounds checking
	 * throw index_out_of_bound if pos is not in [0, size)
	 */
	T & at(const size_t &pos)
	{
		if(pos < 0 || pos >= NowLength)
			throw index_out_of_bound();
		return storage[pos];
	}
	const T & at(const size_t &pos) const
	{
		if(pos < 0 || pos >= NowLength)
			throw index_out_of_bound();
		return storage[pos];
	}
	/**
	 * assigns specified element with bounds checking
	 * throw index_out_of_bound if pos is not in [0, size)
	 * !!! Pay attentions
	 *   In STL this operator does not check the boundary but I want you to do.
	 */
	T & operator[](const size_t &pos)
	{
		if(NowLength == 0)
			throw container_is_empty();
		return storage[pos];
	}
	const T & operator[](const size_t &pos) const
	{
		if(NowLength == 0)
			throw container_is_empty();
		return storage[pos];
	}
	/**
	 * access the first element.
	 * throw container_is_empty if size == 0
	 */
	const T & front() const
	{
		if(NowLength == 0)
			throw container_is_empty();
		return storage[0];
	}
	/**
	 * access the last element.
	 * throw container_is_empty if size == 0
	 */
	const T & back() const
	{
		if(NowLength == 0)
			throw container_is_empty();
		return storage[NowLength - 1];
	}
	/**
	 * returns an iterator to the beginning.
	 */
	iterator begin()
	{
		return iterator(0, this);
	}
	const_iterator cbegin() const
	{
		return const_iterator(0, this);
	}
	/**
	 * returns an iterator to the end.
	 */
	iterator end()
	{
		return iterator(NowLength, this);
	}
	const_iterator cend() const
	{
		return const_iterator(NowLength, this);
	}
	/**
	 * checks whether the container is empty
	 */
	bool empty() const
	{
		return (NowLength == 0);
	}
	/**
	 * returns the number of elements
	 */
	size_t size() const
	{
		return NowLength;
	}
	/**
	 * returns the number of elements that can be held in currently allocated storage.
	 */
	size_t capacity() const
	{
		return MaxSize;
	}
	/**
	 * clears the contents
	 */
	void clear()
	{
		NowLength = 0;
	}
	/**
	 * inserts value before pos
	 * returns an iterator pointing to the inserted value.
	 */
	iterator insert(iterator pos, const T &value)
	{
		if(NowLength == MaxSize - 1) DoubleSpace();
		new(storage + NowLength) T(storage[NowLength - 1]);
		for(iterator iter = end() - 1; iter != pos; --iter)
		{
			*iter = *(iter - 1);
		}
		++NowLength;
		*pos = value;
		return pos;
	}
	/**
	 * inserts value at index ind.
	 * after inserting, this->at(ind) == value is true
	 * returns an iterator pointing to the inserted value.
	 * throw index_out_of_bound if ind > size (in this situation ind can be size because after inserting the size will increase 1.)
	 */
	iterator insert(const size_t &ind, const T &value)
	{
		if(NowLength == MaxSize) DoubleSpace();
		if(ind > NowLength) throw index_out_of_bound();
		new(storage + NowLength) T(storage[NowLength - 1]);
		for(iterator iter = end() - 1; iter != iterator(ind, this); --iter)
		{
			*iter = *(iter - 1);
		}
		++NowLength;

		//new (storage + int) T(value);
		storage[ind] = value;
		return iterator(ind, this);
	}
	/**
	 * removes the element at pos.
	 * return an iterator pointing to the following element.
	 * If the iterator pos refers the last element, the end() iterator is returned.
	 */
	iterator erase(iterator pos)
	{
		if(*pos == NowLength - 1)
		{
			--NowLength;
			return end();
		}
		for(iterator iter = pos; iter != end() - 1; ++iter)
		{
			*iter = *(iter + 1);
		}
		--NowLength;
		return pos;
	}
	/**
	 * removes the element with index ind.
	 * return an iterator pointing to the following element.
	 * throw index_out_of_bound if ind >= size
	 */
	iterator erase(const size_t &ind)
	{
		if(ind == NowLength - 1)
		{
			--NowLength;
			return end();
		}
		for(iterator iter = iterator(ind, this); iter != end() - 1; ++iter)
		{
			*iter = *(iter + 1);
		}
		--NowLength;
		return iterator(ind, this);
	}
	/**
	 * adds an element to the end.
	 */
	void push_back(const T &value)
	{
		if(NowLength == MaxSize - 1)
			DoubleSpace();
        new(storage + NowLength) T(value);
		NowLength++;
	}
	/**
	 * remove the last element from the end.
	 * throw container_is_empty if size() == 0
	 */
	void pop_back()
	{
		if(NowLength == 0) throw container_is_empty();
		--NowLength;
	}
	friend QDataStream& operator >> (QDataStream &in, vector &rhs) {
		int tmp;
		in >> tmp;
		rhs.clear();
		for (int i = 0; i < tmp; ++i) {
            rhs.push_back(T());
            in >> rhs.storage[rhs.NowLength - 1];
		}
		return in;
	}
	friend QDataStream& operator << (QDataStream &out, const vector &rhs) {
		out << rhs.NowLength;
		for (int i = 0; i < rhs.NowLength; ++i)
			out << rhs.storage[i];
		return out;
	}
};

}


#endif
