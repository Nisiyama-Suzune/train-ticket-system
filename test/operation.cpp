#include "vector.hpp"
#include <QDir>
#include <qfile.h>
#include <iostream>
#include <qstringlist.h>
#include <qtextstream.h>
#include <qstring.h>

struct buy_data
{
	QString name;
	int ID;
	QString operation;
	int num;
	QString kind_of_seat;
	QString train_ID;
	QString from_station;
	QString to_station;
	int date;
	friend QTextStream& operator << (QTextStream& out, const buy_data& x) {
		out << "name         = " << x.name << endl;
		out << "ID           = " << x.ID << endl;
		out << "operation    = " << x.operation << endl;
		out << "num          = " << x.num << endl;
		out << "kind_of_seat = " << x.kind_of_seat << endl;
		out << "train_ID     = " << x.train_ID << endl;
		out << "from_station = " << x.from_station << endl;
		out << "to_station   = " << x.to_station << endl;
		out << "date         = " << x.date << endl;
		return out;
	}
};

buy_data operation_transform(QString str)
{
	QTextStream sin(&str);
	QTextStream cout(stdout);

	buy_data ans;
	QString tmp = sin.readLine();
	QStringList parts = tmp.split(' ');

	ans.name = parts[0];

	QTextStream cin1(&parts[1]);
//	ans.ID = parts[1];
	cin1 >> ans.ID;

	ans.operation = parts[2];

	QTextStream cin2(&parts[3]);
	cin2 >> ans.num;

	ans.kind_of_seat = parts[4];
	ans.train_ID = parts[7];
	ans.from_station = parts[9];
	ans.to_station = parts[11];

	QStringList date_parts = parts[13].split('-');
	int date = 0;
	for (int i = 0; i < 4; ++i)
		date = date * 10 + (date_parts[0][i].toLatin1() - '0');
	date = date * 100;
	if (date_parts[1][1].toLatin1() >= '0' && date_parts[1][1].toLatin1() <= '9') {
		date += 10 * (date_parts[1][0].toLatin1() - '0');
		date += date_parts[1][1].toLatin1() - '0';
	} else date += date_parts[1][0].toLatin1() - '0';
	date *= 100;
	if (date_parts[2][1].toLatin1() >= '0' && date_parts[2][1].toLatin1() <= '9') {
		date += 10 * (date_parts[2][0].toLatin1() - '0');
		date += date_parts[2][1].toLatin1() - '0';
	} else date += date_parts[2][0].toLatin1() - '0';
	ans.date = date;
	cout << ans << endl;
	return ans;
}
void test4()
{
	QString directory = QDir::currentPath();
	QTextStream cout(stdout);
	cout << directory << endl;
	directory += "/../load/operation.out";
	QFile file(directory);
//	QFile file("/home/yanhongyu/Git/Data_structure/load/operation.out");
	if (!file.open(QIODevice::ReadOnly|QIODevice::Text)) {
		std::cout << "open file fail" << std::endl;
		return ;
	}
	QTextStream fin(&file);
//	QTextStream cout(stdout);

	QString ans;
	while (fin.readLineInto(&ans)) {
		buy_data buy_ans = operation_transform(ans);
	}
}
