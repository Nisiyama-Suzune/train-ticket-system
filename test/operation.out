鲁技 047066000 bought 16 一等座 tickets of G8903 from 古田北 to 福州 in 2017-4-4
方卓思 062183201 bought 7 二等座 tickets of G4907 from 合肥南 to 安庆 in 2017-3-29
卢郝 064222000 bought 5 硬座 tickets of D7261 from 中山 to 珠海 in 2017-4-2
葛派 042040000 bought 10 无座 tickets of D7833 from 齐齐哈尔南 to 齐齐哈尔 in 2017-3-30
吴征 005050000 bought 1 一等座 ticket of C7103 from 广州 to 东莞 in 2017-4-1
窦欧 037077000 bought 16 二等座 tickets of D9 from 北京 to 秦皇岛 in 2017-4-22
韦许祖 048188001 bought 4 无座 tickets of K5171 from 牡丹江 to 绥芬河 in 2017-4-17
凤听 052157000 bought 16 二等座 tickets of G4823/G4826 from 麻城北 to 南京南 in 2017-4-14
华拥 026094000 bought 19 硬座 tickets of Z36 from 郑州 to 北京西 in 2017-3-31
魏毕 028029000 bought 17 二等座 tickets of C7725 from 小榄 to 新会 in 2017-4-4
任冰 056031000 bought 19 二等座 tickets of D6054 from 青州市 to 济南 in 2017-4-5
陈智 009204000 bought 11 无座 tickets of G172 from 沧州西 to 天津西 in 2017-3-30
鲁压 047019000 bought 14 一等座 tickets of G2915 from 肇庆东 to 深圳北 in 2017-4-8
范忆 044075000 bought 6 特等座 tickets of G1233/G1236 from 泰安 to 滨海 in 2017-4-26
鲁鞋既 047125032 bought 9 无座 tickets of D3231 from 雁荡山 to 鳌江 in 2017-4-13
钱划窗 001153004 bought 2 一等座 tickets of G7317 from 嘉兴南 to 桐乡 in 2017-4-16
方罪佳 062123190 bought 13 无座 tickets of K4089 from 惠州 to 东莞东 in 2017-4-23
郝藸 073142000 bought 14 硬座 tickets of T3089 from 天柱山 to 兴国 in 2017-4-9
陶典盖 029165144 bought 14 特等座 tickets of C1246 from 吉林 to 长春 in 2017-4-15
孔洗萨 023025151 bought 11 一等座 tickets of G476 from 济南 to 济南西 in 2017-4-20
